scality/backbeat
# backport: S3C-1640 CRR retry feature

Depends on https://github.com/scality/backbeat/pull/406

backport: S3C-1640 CRR retry feature

backport the following commits from development/8.1:

* ec16f1f4 FT: Add failure retry routes
* e74e871f FT: Add retry support using kafka and redis
* bb45ac55 rf: ZENKO-147 Refactor retry CRR API
* 43402552 bf: ZENKO-147 Check site when setting failed hash
* cdb417b3 ft: ZENKO-147 delete hash field on retry
* 45c9d419 ft: ZENKO-147 Make CRR retry configurable
* 7d06d0ea ft: S3C-1398 Use retry CRR topic
* 606c0f54 improvement: ZENKO-353 Avoid storing all objectMD
* 2da08f9e ft: ZENKO-147 Use Redis keys instead of hash
* a145a16b rf: ZENKO-143 move attachReqUids() helper
* 3abaaaed bugfix: ZENKO-1144 Use sorted sets for failures
* 7c31e958 bugfix: ZENKO-1144 fix edge cases
* b4568f6a bugfix: ZENKO-1144: Improve listing response time
* 84b95f72 bugfix: ZENKO-1024 Enhance functional tests

partial backport of the following commit from development/8.0:

* 0db307b9 ft: add site-level metrics

Note: only the changes necessary to use the "_redisClient" field in
BackbeatAPI.js have been backported from 0db307b9.

Manual test done:

* shutdown target S3C site
* put a small object on a replicated bucket
* wait 5mn until status becomes FAILED
* do a curl request on http://localhost:8900/_/crr/failed: the FAILED entry appears in the result
* restart target S3C site
* do a POST request on the same URL: the entry becomes PENDING
* the entry replicates and status becomes SUCCESS
